Fixes #20607 - Allow hiding fields which are missing from api output

ofedoren opened the pull request:
Related to https://github.com/theforeman/hammer-cli-foreman/pull/345

mbacovsky commented:
[test]

tstrachota commented:
[test]

mbacovsky commented:
[test]

ofedoren commented:
@mbacovsky  , @tstrachota updated.

tstrachota commented:
[test]

tstrachota commented on lib/hammer_cli/output/adapter/abstract.rb:
```diff
@@ -55,13 +55,24 @@ def self.data_for_field(field, record)
     end
 
     def data_for_field(field, record)
-      Abstract.data_for_field(field, record)
```
Btw you can use rescue without the additional begin/end block:
```ruby
def data_for_field(field, record)
  Abstract.data_for_field(field, record)
rescue HammerCLI::MissingAPIError
  # ...
end
```

tstrachota commented on lib/hammer_cli/output/adapter/abstract.rb:
```diff
@@ -55,13 +55,24 @@ def self.data_for_field(field, record)
+        return Abstract.data_for_field(field, record)
+      rescue HammerCLI::MissingAPIError
+        if field.hide_missing? && (self.class != CSValues)
+          raise HammerCLI::MissingAPIError
```
`rescue HammerCLI::MissingAPIError => e` and then `raise e`. Btw why is `#dafa_for_field` implemented differently from `.data_for_field` ? Shouldn't they behave the same?

tstrachota commented on lib/hammer_cli/exceptions.rb:
```diff
@@ -5,5 +5,6 @@ class OperationNotSupportedError < StandardError; end
   class ModuleLoadingError < StandardError; end
   class ModuleDisabledButRequired < StandardError; end
   class NoCACertificate < StandardError; end
+  class MissingAPIError < StandardError; end
```
The error name is a bit misleading. Could you rename it to something like `DataKeyMissingError`?

tstrachota commented on lib/hammer_cli/output/adapter/csv.rb:
```diff
@@ -83,7 +83,11 @@ def self.expand_container(field, data, formatters)
+        begin
+          return HammerCLI::Output::Adapter::CSValues.data_for_field(field, data)
+        rescue HammerCLI::MissingAPIError
+          return nil
```
Same here. Why is the class implementation different from the instance one?

ofedoren commented:
@tstrachota, updated. 

Thank you for suggesting a better solution. https://github.com/theforeman/hammer-cli-foreman/pull/345 is redundant now and should be closed I guess.

tstrachota commented on lib/hammer_cli/output/adapter/abstract.rb:
```diff
@@ -69,6 +70,13 @@ def output_stream
       $stdout
     end
 
+    def displayable_fields(fields, record)
```
Nice, thank you for extracting it to a method.

I know it's not your code, but since you're touching the area... I noticed that there's some redundancy in the **base adapter**:
https://github.com/theforeman/hammer-cli/blob/master/lib/hammer_cli/output/adapter/base.rb#L31-L34
https://github.com/theforeman/hammer-cli/blob/master/lib/hammer_cli/output/adapter/base.rb#L47
and in the **tree adapter**:
https://github.com/theforeman/hammer-cli/blob/master/lib/hammer_cli/output/adapter/tree_structure.rb#L23-L26
https://github.com/theforeman/hammer-cli/blob/master/lib/hammer_cli/output/adapter/tree_structure.rb#L34

Both adapters first filter out fields that aren't displayable and checks for the same thing later again. I think that we could either use your new method to filter them out once for good in `filter_fields` or remove that `reject` block from filter_fields. The latter is more effective because we wouldn't call `data_for_field` twice.
What do you think?

tstrachota replied:
> And let the field_filter decide which fields will be kept for rendering.

Ack, I don't care that much about which of the ways you choose.

> Also I'm not sure now if displayable_fields method should be in adapter context instead of fields filter.

I was thinking about the same. In future, we'll most likely need to detect what fields are requested to be displayed prior to fetching data from API to conditionally load more/less data (eg. when GrephQL is used). On the other hand the "data for field" aren't known prior to the API request so we'll have to split the logic.
Anyway, I think that we should do this refactoring in a separate PR. This patch is nice and simple improvement that doesn't block us from doing it later. So, let's go step by step and use your `displayable_fields` for now.

tstrachota commented on lib/hammer_cli/output/adapter/table.rb:
```diff
@@ -52,7 +53,7 @@ def print_collection(all_fields, collection)
       output_stream.puts line unless formatted_collection.empty? || @context[:no_headers]
 
-      if collection.meta.pagination_set? && collection.count < collection.meta.subtotal
+      if collection.respond_to?(:meta) && collection.meta.pagination_set? && collection.count < collection.meta.subtotal
```
Thank you for fixing this!

tstrachota commented on lib/hammer_cli/output/fields.rb:
```diff
@@ -18,11 +19,15 @@ def hide_blank?
       @hide_blank
     end
 
+    def hide_missing?
```
Nitpick: could be `attr_reader :hide_missing?`

ofedoren commented:
@tstrachota, updated.

Since ```displayable_fields``` is used in **csv** and **table** adapters, I refactored other adapters in the same manner as well.

Also, because [hide_missing?](https://github.com/ofedoren/hammer-cli/blob/b209d202a37b63deaecbe8ae1f69cab348aec0aa/lib/hammer_cli/output/fields.rb#L22) returns a boolean value, I'd like to keep as it is, since making ```attr_reader``` with question mark is not allowed in ruby.

tstrachota commented on lib/hammer_cli/output/adapter/base.rb:
```diff
@@ -13,6 +13,8 @@ def print_record(fields, record)
 
     def print_collection(fields, collection)
+      fields = field_filter.filter(fields)
+      fields = displayable_fields(fields, collection.first)
```
It was missing on purpose. This functionality is disabled for table and csv adapters because we can't tell what data are in the other records. Using the first record for hiding fields that are missing would be probably fair guess, but the same method can remove also fields that are blank (`""`). That could result in some false positives -> whole column would be hidden just because the first record contained empty string.

tstrachota commented on lib/hammer_cli/output/adapter/csv.rb:
```diff
@@ -157,6 +157,7 @@ def print_record(fields, record)
     end
 
     def print_collection(fields, collection)
+      fields = displayable_fields(fields, collection.first)
```
Similarly to my note in the base adapter. We shouldn't filter fields in printing collections in csv.

tstrachota commented on lib/hammer_cli/output/adapter/tree_structure.rb:
```diff
@@ -7,33 +7,28 @@ def initialize(context={}, formatters={})
-
-      fields.reduce({}) do |hash, field|
+      rendered = {}
+      fields.each do |field|
```
What was the reason for translating reduce into more verbose each block?

ofedoren commented:
@tstrachota, updated.

I changed ```displayable_fields``` method, so **csv** and **table** adapters will not display field only when data is missing for this field in the first record (it does not remove field if data is just blank and I assume if data is missing in the first record, then it's missing in the others as well).

tstrachota commented:
Works as expected :champagne: . Thank you @ofedoren !
